cmake/FindLibopencm3.cmake

```diff
@@ -0,0 +1,177 @@
+endif()
+
+get_filename_component(Libopencm3_ROOT_DIR "${CMAKE_CURRENT_LIST_DIR}/.." ABSOLUTE)
+
+set(LOCM3_DATA_FILE ${Libopencm3_ROOT_DIR}/ld/devices.data)
+
+
+if (NOT EXISTS ${LOCM3_DATA_FILE})
+	message(FATAL_ERROR "Unable to find device data file ${LOCM3_DATA_FILE}!")
+endif()
+
+set(Libopencm3_LINKER_SCRIPT ${CMAKE_BINARY_DIR}/gen.${DEVICE}.ld)
+
+function(_genlink_obtain DEVICE PROPERTY OUTPUT)
+	execute_process(COMMAND 
+		${PYTHON_EXE} ${Libopencm3_ROOT_DIR}/scripts/genlink.py ${LOCM3_DATA_FILE} ${DEVICE} ${PROPERTY}
+		OUTPUT_VARIABLE OUT_DATA
+		RESULT_VARIABLE SUCCESS
+		)
+	if ("${SUCCESS}" EQUAL "0")
+		message(DEBUG ">> ${OUT_DATA}")
+		set("${OUTPUT}" "${OUT_DATA}" PARENT_SCOPE)
+	else()
+		message(FATAL_ERROR "Unable to obtain ${PROPERTY} for device ${DEVICE}!")
+	endif()
+endfunction()
+
+_genlink_obtain(${DEVICE} FAMILY DEVICE_FAMILY)
+_genlink_obtain(${DEVICE} SUBFAMILY DEVICE_SUBFAMILY)
+_genlink_obtain(${DEVICE} CPU DEVICE_CPU)
+_genlink_obtain(${DEVICE} FPU DEVICE_FPU)
+_genlink_obtain(${DEVICE} CPPFLAGS Libopencm3_DEFINITIONS)
+_genlink_obtain(${DEVICE} DEFS DEVICE_DEFS)
+
+if ("${DEVICE_FAMILY}" STREQUAL "")
+	message(FATAL_ERROR "${DEVICE} not found in ${LOCM3_DATA_FILE}")
+endif()
+
+string(REPLACE " " ";" DEVICE_DEFS ${DEVICE_DEFS})
+
+# This essentially duplicates actions of genlink-config.mk
+set(LOCM3_THUMB_DEVS 
+	cortex-m0
+	cortex-m0plus
+	cortex-m3
+	cortex-m4
+	cortex-m7
+	)
+
+set(ARCH_FLAGS "")
+list(APPEND ARCH_FLAGS -mcpu=${DEVICE_CPU})
+
+if ("${DEVICE_CPU}" IN_LIST LOCM3_THUMB_DEVS)
+	list(APPEND ARCH_FLAGS -mthumb)
+endif()
+
+if ("${DEVICE_FPU}" STREQUAL "soft")
+	list(APPEND ARCH_FLAGS -msoft-float)
+elseif("${DEVICE_FPU}" STREQUAL "hard-fpv4-sp-d16")
+	list(APPEND ARCH_FLAGS -mfloat-abi=hard -mfpu=fpv4-sp-d16)
+elseif("${DEVICE_FPU}" STREQUAL "hard-fpv5-sp-d16")
+	list(APPEND ARCH_FLAGS -mfloat-abi=hard -mfpu=fpv5-sp-d16)
+else()
+	message(FATAL_ERROR "No match for the FPU flags")
+endif()
+
+# Find library for given family or subfamily
+set(OPENCM3_LIBNAME
+	opencm3_${DEVICE_FAMILY}
+	opencm3_${DEVICE_SUBFAMILY}
+	)
+
+foreach(LOCM3_CANDIDATE ${OPENCM3_LIBNAME})
+	if (EXISTS ${Libopencm3_ROOT_DIR}/lib/lib${LOCM3_CANDIDATE}.a)
+		# If found, then generate the linker script
+		execute_process(COMMAND ${CMAKE_C_COMPILER} 
+			${ARCH_FLAGS} ${DEVICE_DEFS} 
+			-P -E ${Libopencm3_ROOT_DIR}/ld/linker.ld.S 
+			-o ${Libopencm3_LINKER_SCRIPT}
+			RESULT_VARIABLE CPP_RESULT
+			)
+
+		if (NOT "${CPP_RESULT}" EQUAL "0")
+			message(FATAL_ERROR "Unable to generate linker script for device ${DEVICE}")
+		endif()
+
+		# Provide exported variables
+		set(Libopencm3_LIBRARY ${LOCM3_CANDIDATE})
+		set(Libopencm3_LIBRARY_DIRS ${Libopencm3_ROOT_DIR}/lib)
+		set(Libopencm3_INCLUDE_DIRS ${Libopencm3_ROOT_DIR}/include)
+		set(Libopencm3_DEFINITIONS ${Libopencm3_DEFINITIONS} ${ARCH_FLAGS})
+		set(Libopencm3_LINK_OPTIONS -static -nostartfiles ${ARCH_FLAGS} -T${Libopencm3_LINKER_SCRIPT})
```

Comment:
Here I deliberately add -static and -nostartfiles. Genlink-config.mk is not doing this, but I am not quite sure if there are any valid use cases, where you might not want these two switches. Having them here will prevent from nasty ugly messages from compiler, if used without additional flags in CMakeLists.txt.